Add Renesas H3ULCB board

julien-massot opened the pull request:
This pull request adds a very minimal support in order to run Zephyr on the Cortex-r7 
present on Renesas H3ULCB board.

H3ULCB have cortex A5x Cores and a 'Realtime processor' based on Cortex-R7.

With this minimal support it is possible to start the hello world, blinky and button demo.
The output is printed on the ram console.

Drivers such as Pinmux, UART, CAN will come later and are already tested in our feature branches
such as 'renesas' branch of https://github.com/iotbzh/zephyr.

This PR create also a renesas_rcar soc family and a 'gen3' soc_series, the different SoCs of the Gen3 series will be
able to reuse this work with just few dts entries and pinmux configuration.

For user which want to talk about Renesas R-Car support within Zephyr, feel free to join the new shiny 'renesas' channel on
Zephyr slack.

Commit:
soc: arm: add Renesas rcar_gen3 series support

Most of the Renesas RCar Gen3 based SoC contains a Cortex R7
processor.
This processor has access to the same memory mapped devices than
the Cortex-A5x cores.

- CPU operates upto 800MHz
- Can use ram area from 0x40040000 to 0x42000000
- Has 512 interrupts on GIC-400 compliant with Arm GICv2

Add support for r8a77951 as first SoC of this series which is also
known as H3 ES2.0 and is present present on different boards such as
Salvator and R-Car Starter Kit(H3ulcb).

This first SoC definition is just enough to print Hello World in a
ram console.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
boards: arm: Add support for R-Car H3ULCB board

Add basic configuration for H3ULCB, just enough to see the
Zephyr boot banner on the ram console.

This configuration make use of the Cortex-R7 present on
r8a977951 SoC.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

aaillet committed:
doc: add documentation for Renesas rcar_h3ulcb board

Adding ulcb boards documentation based on Renesas official documentation
and following zephyr guideline.

The documentation is describing the board and the current
Zephyr support.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

Commit:
drivers: clock_control: add R-Car CPG MSSR driver

Clock Pulse Generator, Module Standby Software Reset, are registers
presents in Renesas Gen3 SoC series.

MSSR is used to supply clock to the different modules, shuch as timer,
or UART, it's also possible to issue a reset the different module.

CPG registers allow to get the rate or to set some divider like for
the CAN clock.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
dts: arm: rcar_gen3_cr7: add cmt timer and clock controller

The Compare match timer can be found on Renesas
RCar Gen3 soc series.

It depends on clock controller to supply clock to the
CMT module.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
dts: bindings: gpio: add definition for Renesas RCar gpio

Renesas RCar gpio controller can manage up to 32 gpio
per bank.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
dts: arm: rcar: add GPIO banks 5 and 6

Renesas RCar Gen3 series have up to 8 GPIOs
bank.

Add bank 5 and bank 6, that is used to manage user led and
switches on different demo board.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
boards: arm: rcar_h3ulcb: add led0 and sw0

Enable the user led and user switch that can be found
on h3ulcb, they are both connected to the GPIO 6 bank.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
boards: arm: rcar_h3ulcb: add Timer support

Enable CMT timer that can be found on H3ULCB board.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

aaillet committed:
boards: doc: Update documentation for R-Car H3ULCB

Update supported features for Renesas
R-Car H3ULCB board.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

Commit:
dts: bindings: clock: add Renesas clock control

This add bindings for Renesas CPG, MSSR clock control
driver.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
include: Add Renesas clock control driver

DTS bindings file for Renesas RCar CPG clock control.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
dts: bindings: add binding for Renesas RCar CMT timer

Add device tree binding for Compare Match Timer that can
be found on various Renesas RCar SoC.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
test: kernel: context: Add Renesas RCar CMT timer

This test require explicit definition of the timer irq.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

julien-massot commented:
Finally I decide to make this PR bigger than the first proposal,
by adding clock control, timer and gpio driver.

Please let us know what is the preferred strategy.

julien-massot commented on soc/arm/renesas_rcar/gen3/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,17 @@
+	default "gen3"
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 800000000
```
this value is wrong should be 32KHz.

ioannisg commented:
@stephanosio you might want to take a look as well.

carlocaione commented:
Hey @julien-massot a general comment: can you fix the define indentation madness in this patchset? ;)

Also a general question (maybe for @ioannisg @nashif and others) should the the device MMIO API enforced for all the new drivers? Or is that only nice-to-have?

julien-massot commented:
> Hey @julien-massot a general comment: can you fix the define indentation madness in this patchset? ;)
Hi @carlocaione Indeed I did read this code so much so that I didn't notice anymore that's indentation is odd,
will look at it again.

Thanks for the review.
I'm also interested at any hint which can make the review easier, one way could be to send each driver: timer, gpio and clock alone, and then the SoC support so that each patchset may be easier to handle.

carlocaione commented:
> I'm also interested at any hint which can make the review easier, one way could be to send each driver: timer, gpio and clock alone, and then the SoC support so that each patchset may be easier to handle.

My take on this is (and other devs can have a different opinion on this): the minimum set of things to make the CI happy when testing the pushed code.

So for example I wouldn't push the GPIO driver alone when there is no way for the CI to test the code because there is no soc support for that. So I would push the soc / board enabling code and the minimum set of drivers to make it testable on the CI.

julien-massot commented:
Ok so I will keep the current strategy. That's already the minimum.

julien-massot commented:
Just rebased:
- Reindent the #define
- Fix SYS_CLOCK_HW_CYCLES_PER_SEC to 32KHz and move it to board defconfig
- Rebased on top of master

mnkp commented on drivers/gpio/gpio_rcar.c:
```diff
@@ -0,0 +1,334 @@
+	sys_slist_t cb;
+};
+
+#define IOINTSEL 0x00   /* General IO/Interrupt Switching Register */
+#define INOUTSEL 0x04   /* General Input/Output Switching Register */
+#define OUTDT    0x08   /* General Output Register */
+#define INDT     0x0c   /* General Input Register */
+#define INTDT    0x10   /* Interrupt Display Register */
+#define INTCLR   0x14   /* Interrupt Clear Register */
+#define INTMSK   0x18   /* Interrupt Mask Register */
+#define MSKCLR   0x1c   /* Interrupt Mask Clear Register */
+#define POSNEG   0x20   /* Positive/Negative Logic Select Register */
+#define EDGLEVEL 0x24   /* Edge/level Select Register */
+#define FILONOFF 0x28   /* Chattering Prevention On/Off Register */
+#define OUTDTSEL 0x40   /* Output Data Select Register */
+#define BOTHEDGE 0x4c   /* One Edge/Both Edge Select Register */
```
These values should ideally be provided by a vendor header file from a board support package. We keep these HAL support files in `../modules/hal/`. The name of the driver is `gpio_rcar` and that implies it should support all Renesans "R-Car" family SoCs. I'm not familiar with "R-Car" however typically register addresses change between SoC series. The one defined in this driver are tailored for R-Car-H3 SoC.

julien-massot replied:
On RCar SoC these registers are almost compatible for R-Car GEN 1,2,3.
From what I see in Linux If one want to add support for Gen1 and Gen2, only the two last registers 
OUTDTSEL and BOTHEDGE may not be available.

mnkp replied:
As per #33171 Renesas doesn't provide its own HAL, we can keep register defines directly in the driver.

julien-massot commented:
Hi @ioannisg does it's possible to have review on this patchset ?
Thanks

julien-massot commented on drivers/gpio/gpio_rcar.c:
```diff
@@ -0,0 +1,334 @@
+		.reg_addr = DT_INST_REG_ADDR(n),		    \
+		.reg_size = DT_INST_REG_SIZE(n),		    \
+		.init_func = gpio_rcar_##n##_init,		    \
+		.clock_controller = DT_INST_CLOCKS_LABEL(n),	    \
```
will change to use DT_INST_CLOCKS_CTLR

julien-massot commented on drivers/timer/rcar_cmt_timer.c:
```diff
@@ -0,0 +1,141 @@
+#define TIMER_BASE_ADDR        DT_INST_REG_ADDR(0)
+#define TIMER_CLOCK_FREQUENCY  DT_INST_PROP(0, clock_frequency)
+
+#define CLOCK_CONTROLLER       DT_INST_CLOCKS_LABEL(0)
```
will change to use DT_INST_CLOCKS_CTLR

julien-massot commented:
Just rebased and convert GPIO and Timer driver to use DEVICE_DT_GET instead of DT_INST_CLOCKS_LABEL.

ioannisg commented on boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml:
```diff
@@ -0,0 +1,8 @@
+toolchain:
+  - zephyr
+  - gnuarmemb
+ram: 768
```
How about flash @julien-massot ?

julien-massot replied:
There is no flash support at the moment. The cortex-r7 is a remote processor of the CA57/CA53, we only use ram area at the moment, the CA57/CA53 processors load the firmware and then start this CR7 processor.

ioannisg reviewed (approved):
Looks good to me.
@julien-massot I understand that this builds with XIP=n, correct?

ioannisg commented:
@julien-massot I realize (also by looking at the comment by @mnkp ) that we're not having a HAL module for this Renesas platform. Is this possible to actually provide one? Then all hw register definitions would be taken from that module repository, see how is this done for all other ARM vendor SoC support.

julien-massot commented:
> Looks good to me.
> @julien-massot I understand that this builds with XIP=n, correct?
@ioannisg Yes, your understanding is correct.

julien-massot commented:
> @julien-massot I realize (also by looking at the comment by @mnkp ) that we're not having a HAL module for this Renesas platform. Is this possible to actually provide one? Then all hw register definitions would be taken from that module repository, see how is this done for all other ARM vendor SoC support.

Indeed I didn't add a HAL layer yet, I didn't had the feeling that a HAL was useful at this point.
How do you want to proceed about this PR ?
Does we go for the implementation as is, and introduce a hal layer at the same time than introducing
pinmux and other driver such as UART, CAN, I2C or IPM.
Or does we introduce the hal layer right now and move register definition ?

julien-massot commented:
just rebased to fix trivial conflict on codeowner file

Commit:
drivers: gpio: add Renesas RCar gpio

Add GPIO controller driver that can be found on Renesas
RCar gen3 soc series.

Controller can handle up to 32 GPIOs per banks.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

julien-massot commented:
trivial rebase again

carlescufi commented:
@galak any objections to this?
@mnkp can you take another look?

carlescufi commented on boards/arm/rcar_h3ulcb/doc/rcar_h3ulcb.rst:
```diff
@@ -0,0 +1,169 @@
+
+.. code-block:: bash
+
+	git clone --branch renesas https://github.com/iotbzh/openocd.git
```
Maybe you could consider upstreaming those patches or merging them into [our OpenOCD fork](https://github.com/zephyrproject-rtos/openocd), which includes support for Zephyr thread awareness. Otherwise users of this board won't get those extra features.

julien-massot replied:
Hi, Zephyr awareness is included in my own fork, i'm still thinking about either cleaning and contributing to zephyr openocd, or upstream openocd, or finding another tool.

carlescufi reviewed (approved):
LGTM in general. The OpenOCD integration can be improved later.

stephanosio commented on drivers/clock_control/clock_control_rcar_cpg_mssr.c:
```diff
@@ -0,0 +1,226 @@
+
+	addr = config->base_address + CANFDCKCR;
+	if (enable) {
+		if (CANFDCKCR_PARENT_CLK_RATE % rate) {
```
Nit: implicit int-bool conversion/interpretation

julien-massot replied:
will repalce by  (CANFDCKCR_PARENT_CLK_RATE % rate) != 0

julien-massot replied:
fixed

stephanosio commented on drivers/clock_control/clock_control_rcar_cpg_mssr.c:
```diff
@@ -0,0 +1,226 @@
+{
+	const struct rcar_mssr_config *config = DEV_CFG(dev);
+	uint32_t addr, divider;
+	int key, ret;
```
`irq_lock` returns `unsigned int`.

julien-massot replied:
thanks will change

stephanosio commented on drivers/clock_control/clock_control_rcar_cpg_mssr.c:
```diff
@@ -0,0 +1,226 @@
+				     uint32_t module, bool enable)
+{
+	const struct rcar_mssr_config *config = DEV_CFG(dev);
+	uint32_t reg = module / 100;
+	uint32_t bit = module % 100;
+	uint32_t bitmask = BIT(bit);
```
This seems somewhat unusual. Just double-checking that `module` never exceeds 31 for every 100 interval (i.e. 0-31, 100-131, 200-231, ...).

julien-massot replied:
ok will add a check

julien-massot replied:
fixed

stephanosio commented on drivers/clock_control/clock_control_rcar_cpg_mssr.c:
```diff
@@ -0,0 +1,226 @@
+	uint32_t bitmask = BIT(bit);
+	uint32_t reg_val;
+
+	int key = irq_lock();
```
`unsigned int`

julien-massot replied:
ok

stephanosio commented on drivers/clock_control/clock_control_rcar_cpg_mssr.c:
```diff
@@ -0,0 +1,226 @@
+
+	int key = irq_lock();
+
+	reg_val = sys_read32(config->base_address + rmstpsr[reg]);
```
A boundary check (at least, `ASSERT`) for `reg` would be nice given the nature of this SoC and its applications.

julien-massot replied:
will add a check like ASSERT (reg < ARRAY_SIZE(rmstpsr))

julien-massot replied:
done

stephanosio commented on drivers/clock_control/clock_control_rcar_cpg_mssr.c:
```diff
@@ -0,0 +1,226 @@
+static int cpg_mssr_blocking_start(const struct device *dev,
+			       clock_control_subsys_t sys)
+{
+	struct rcar_cpg_clk *clk = (struct rcar_cpg_clk *) sys;
```
Nit: spurious space between `)` and `sys`.

julien-massot replied:
ok

stephanosio commented on drivers/clock_control/clock_control_rcar_cpg_mssr.c:
```diff
@@ -0,0 +1,226 @@
+		*rate = S3D4_CLK_RATE;
+		break;
+	default:
+		ret = -ENOTSUP;
+	}
```
Although not required, `break` here would be nice since this is not an intentional fall-through.

p.s. are we not getting compiler warnings about this?

julien-massot replied:
at least with arm-zephyr-eabi-gcc (crosstool-NG 1.24.0.37-3f461da-dirty) 9.2.0 I don't have any warning.
will add break statement

stephanosio commented on drivers/timer/rcar_cmt_timer.c:
```diff
@@ -0,0 +1,137 @@
+/*
```
Just double-checking: have you verified that this passes all the relevant kernel tests in real hardware? (e.g. `tests/kernel/timer`).

julien-massot replied:
These three timer test do pass: 
timer_api
timer_error_case
timer_monotonic

the kernel/timer/starve do not.
have to investigate

stephanosio replied:
Do `kernel/timer/starve` and the rest of the tests pass now?

julien-massot replied:
I finally succeed to run twister with success on another branch, this one lacks openocd support for west and serial output.
I'm running starve test right now, since this one is out of twister default test suite.

julien-massot replied:
kernel/timer/starve does pass

stephanosio commented on drivers/timer/rcar_cmt_timer.c:
```diff
@@ -0,0 +1,137 @@
+	return 0;
+}
+
+uint32_t z_timer_cycle_get_32(void)
+{
+	return cycle_count;
+}
```
Pretty sure the `test_prevent_interrupt` test in `tests/kernel/interrupt` will not pass with this implementation.

julien-massot replied:
Correct: by calling irq_lock,  k_busy_wait never return since Timer ISR is not serviced, and so cycle_count is not incremented.
Well the way this timer is working is that the counter is reset to 0 after reaching the match value, and an overflow bit is set.
So at this point I can only add a maximum of CYCLES_PER_TICK to the cycle_count (which is already better), but should not be enough to unlock the k_busy_wait(MS_TO_US(1000));

stephanosio replied:
One way to address this issue is, if possible, to return the value of the counter register.

julien-massot replied:
Does having k_busy_wait working with irq locked is manadatory ?
In this case I guess this hw timer is not suitable since it just count up to one tick.

stephanosio replied:
> Does having k_busy_wait working with irq locked is manadatory ?

A de facto requirement, one could say: https://github.com/zephyrproject-rtos/zephyr/issues/23519

All tests under `tests/` must pass. Try running a full sanitycheck with the real hardware to see if there are any more failures.

andyross replied:
FWIW: k_busy_wait() needs to work in all contexts, including both locking and high priority ISRs that the timer interrupt can't preempt.  It's used everywhere.  But it doesn't absolutely have to be based on the timer counter, that's just the default.  There are platforms like nRF5 that use a separate mechanism (e.g. another counter device, a cycle-counted loop, etc...).  See CONFIG_ARCH_HAS_CUSTOM_BUSY_WAIT.

But that said: long term you really want a tickless-capable timer driver.

julien-massot replied:
Thanks for your time and all these advises, I will try to think about a solution.
One of my attempt was to use private timer located in the Cortex-r, I finally gave up with this timer
because it is connected to a different GIC controller than other shared peripherals.
Finally it could be suitable for implementing a custom_busy_wait since it doesn't require interrupts.
One other solution could be to use another channel of this cmt timer that could be kept running, while this one can be kept looping on tick value.

@andyross indeed supporting tickless is the long term solution, in best scenario using the generic arm timer,
if I find a solution for managing to GIC controller within Zephyr.

julien-massot replied:
@andyross @stephanosio k_busy_wait issue has been fixed by using a second channel of this timer which is not reset when reaching the tick match value. Tickless support will be added later.

stephanosio reviewed (dismissed):
Some nits and sanity checking

julien-massot reviewed (commented):
Rebased on top of master and did minor change in timer driver:
z_timer_cycle_get_32 -> sys_clock_cycle_get_32
z_clock_announce -> sys_clock_announce
z_clock_driver_init -> sys_clock_driver_init

Also the major change in timer driver is the use of the second channel of this timer,
the second channel is now used to return the cycle count which allow to busy wait in isr
context.

The other requested change has also been addressed.

ioannisg commented:
@stephanosio would you like to take a look in this PR?

julien-massot commented:
@stephanosio Do you have time to check if your requested changes are addressed by this latest version ?

bbolen commented on drivers/clock_control/clock_control_rcar_cpg_mssr.c:
```diff
@@ -0,0 +1,231 @@
+	0x980, 0x984, 0x988, 0x98c,
+};
+
+#define RMSTPSR(i)      rmstpsr[i]
```
This define is unused.

julien-massot replied:
thanks, will use it in next version instead of accessing the array directly.

bbolen commented on boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7_defconfig:
```diff
@@ -1,8 +1,11 @@
 CONFIG_SOC_SERIES_RCAR_GEN3=y
 CONFIG_BOARD_RCAR_H3ULCB_CR7=y
+CONFIG_CLOCK_CONTROL=y
+CONFIG_CLOCK_CONTROL_RCAR_CPG_MSSR=y
```
Some of the other platforms have something like
select CLOCK_CONTROL_RCAR_CPG_MSSR if CLOCK_CONTROL in the soc Kconfig.soc file instead of defining the specific clock control driver in every board defconfig.

julien-massot replied:
indeed it make sense, will fix

bbolen commented on boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7_defconfig:
```diff
@@ -9,3 +9,5 @@ CONFIG_CONSOLE=y
 CONFIG_FLASH_SIZE=0
 CONFIG_FLASH_BASE_ADDRESS=0
+CONFIG_GPIO=y
+CONFIG_GPIO_RCAR=y
```
GPIO_RCAR could be moved to an soc kconfig file so that it doesn't have to be defined in every board.

julien-massot replied:
about this one, I think prefer to keep it at board level. Since I/O are shared between the multiple processors on this platform,
it let the board integrator the choice to enable or not this driver.

bbolen replied:
What I meant by this was that GPIO_RCAR could move to the soc kconfig file, but CONFIG_GPIO would still be enabled at the board level, i.e. it would work like the clock control kconfig from the other change I mentioned.

julien-massot replied:
got it! thanks a lot

julien-massot replied:
hum, perhaps not, let's imagine that you want to drive an I2c gpio expander, then you don't necessary need GPIO_RCAR.

bbolen replied:
fair enough

stephanosio commented on drivers/gpio/gpio_rcar.c:
```diff
@@ -0,0 +1,326 @@
+	pending = gpio_rcar_read(config, INTDT);
+	mask = gpio_rcar_read(config, INTMSK);
+
+	while ((pending = gpio_rcar_read(config, INTDT) &
+			  gpio_rcar_read(config, INTMSK))) {
+		fsb = find_lsb_set(pending) - 1;
+		gpio_rcar_write(config, INTCLR, BIT(fsb));
```
Should it `INTCLR` after all the callbacks are processed in order to prevent nested entry into the `gpio_rcar_port_isr` and the callbacks?

julien-massot replied:
Indeed that's better, I saw this scheme on other driver but interrupt handling may be different.

julien-massot replied:
fixed

stephanosio commented on drivers/timer/rcar_cmt_timer.c:
```diff
@@ -0,0 +1,151 @@
+	}
+
+	/* Supply clock for both channels */
+	sys_write32(CLKEN0|CLKEN1, TIMER_BASE_ADDR + CMCLKE);
```
Nit: missing space
```suggestion
	sys_write32(CLKEN0 | CLKEN1, TIMER_BASE_ADDR + CMCLKE);
```

julien-massot replied:
ok thanks.

julien-massot replied:
fixed

stephanosio commented on drivers/timer/rcar_cmt_timer.c:
```diff
@@ -0,0 +1,151 @@
+
+	ARG_UNUSED(device);
+	clk = DEVICE_DT_GET(DT_INST_CLOCKS_CTLR(0));
+	if (!clk) {
```
Nit: Implicit type conversion (and MISRA violation)
```suggestion
	if (clk == NULL) {
```

julien-massot replied:
fixed

stephanosio commented on drivers/timer/rcar_cmt_timer.c:
```diff
@@ -0,0 +1,151 @@
+		return -ENODEV;
+	}
+
+	ret = clock_control_on(clk, (clock_control_subsys_t *) &mod_clk);
```
Nit: Spurious space
```suggestion
	ret = clock_control_on(clk, (clock_control_subsys_t *)&mod_clk);
```

julien-massot replied:
fixed

stephanosio commented on drivers/timer/rcar_cmt_timer.c:
```diff
@@ -0,0 +1,151 @@
+		return -ENODEV;
+	}
+
+	ret = clock_control_on(clk, (clock_control_subsys_t *) &mod_clk);
```
Was this cast really necessary? Implicit pointer conversion to `void *` is allowed in C. Are you seeing warnings about this?

julien-massot replied:
Indeed it's allowed and gcc 10.2 doesn't complain without this cast, since other platform (stm32) does this cast, I prefer to keep it.

stephanosio commented on drivers/timer/rcar_cmt_timer.c:
```diff
@@ -0,0 +1,151 @@
+	sys_write32(0, TIMER_BASE_ADDR + CMCNT0_OFFSET);
+
+	for (i = 0; i < 1000; i++) {
+		if (!sys_read32(TIMER_BASE_ADDR + CMCNT0_OFFSET))
+			break;
```
Nit: as per the coding guidelines
```suggestion
		if (!sys_read32(TIMER_BASE_ADDR + CMCNT0_OFFSET)) {
			break;
		}
```

julien-massot replied:
ok

julien-massot replied:
fixed

stephanosio reviewed (commented):
Some more comments

julien-massot commented on drivers/clock_control/clock_control_rcar_cpg_mssr.c:
```diff
@@ -0,0 +1,231 @@
+
+static void cpg_write(uint32_t addr, uint32_t val)
+{
+	sys_write32(~val, CPGWPR);
```
Good catch, about cpg write protect ! surprising that can clock is correctly set in my tree.

julien-massot replied:
fixed

bbolen reviewed (approved):
Aside from stephanosio's Nit comments, looks good to me.

ioannisg commented:
> Aside from stephanosio's Nit comments, looks good to me.

@julien-massot pls, address those, so we can close the review :)

Commit:
drivers: timer: add R-Car cmt driver

Compare Match Timer is a 32 bit compare match timer
that can be found on various Renesas R-Car SoC.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

julien-massot commented:
Thanks to all people who helped me to find some issues and to apply all the Zephyr code guidelines !
